library/general/src/lib/yast2/log_dir_rotator.rb

```diff
@@ -0,0 +1,195 @@
+      end
+    end
+
+    # Write a file in the log directory.
+    #
+    # @param dest [String] basename of destination
+    # @param content [String] content to write
+    def write(dest, content)
+      return if Yast::Mode.test
+      begin
+        File.open(log_dir + "/" + dest, "w") { |file| file.write(content) }
+      rescue SystemCallError => e
+        log.error e.to_s
+      end
+    end
+
+    # Return a suitable name for the devicegraph log directory
+    # depending on the YaST mode (installation / installed system).
+    #
+    # @return [String] directory name with full path
+    def log_dir
+      dir = installation? ? "#{self.class::NAME}-inst" : self.class::NAME
+      base_dir + "/" + dir
+    end
+
+  private
+
+    # Return true if this is some installation mode: installation, update
+    #
+    # @return [Boolean]
+    def installation?
+      Yast::Mode.installation || Yast::Mode.update
+    end
+
+    # Rotate the log directories, depending on current YaST mode:
+    #
+    # During installation (or update or AutoYaST), clear and remove any old
+    # /var/log/YaST2/#{NAME}-inst directory.
+    #
+    # In the installed system, keep a number of old log directories, remove
+    # any older ones in /var/log/YaST2, and rename the ones to keep:
+    #
+    #   rm -rf #{NAME}-03
+    #   mv #{NAME}-02 #{NAME}-03
+    #   mv #{NAME}-01 #{NAME}-02
+    #   mv #{NAME}    #{NAME}-01
+    #
+    # This will NOT create any new log directory.
+    def rotate_log_dirs
+      return unless File.exist?(base_dir)
+      if installation?
+        kill_old_log_dirs([File.basename(log_dir)])
+      else
+        log_dirs = old_log_dirs.sort
+        keep_dirs = log_dirs.shift(self.class::KEEP_OLD_LOG_DIRS)
+        kill_old_log_dirs(log_dirs)
+        keep_dirs.reverse.each { |dir| rename_old_log_dir(dir) }
+      end
+    end
+
+    # Kill (recursively remove) old log directories.
+    #
+    # @param log_dirs [Array<String>] directory names (without path) to remove
+    def kill_old_log_dirs(log_dirs)
+      log_dirs.each do |dir|
+        next unless File.exist?(base_dir + "/" + dir)
+        log.info("Removing old log dir #{dir}")
+        FileUtils.remove_dir(base_dir + "/" + dir)
+      end
+    end
+
+    # Return the old devicegraph log directories for the installed system
+    # currently found in base_dir: ["#{NAME}", "#{NAME}-01", "#{NAME}-02", ...]
+    #
+    # @return [Array<String>] directory names without path
+    def old_log_dirs
+      Dir.entries(base_dir).select do |entry|
+        entry.start_with?(self.class::NAME) && entry != "#{self.class::NAME}-inst"
+      end
+    end
+
+    # Rename an old log directory according to this schema:
+    #
+    #   mv #{NAME}-02 #{NAME}-03
+    #   mv #{NAME}-01 #{NAME}-02
+    #   mv #{NAME}    #{NAME}-01
+    #
+    # @param old_name [String] old directory name (without path)
+    def rename_old_log_dir(old_name)
+      new_name =
+        if old_name =~ /[0-9]+$/
```

Comment:
I suggest something like `/\d+\z/`. In ruby `$` in regexp means matching also endline. So potentianally if filepath contain newline[1] this won't work as intented. and `\d` is shorter then `[0-9]`. More about ruby regexp at https://ruby-doc.org/core-2.5.1/Regexp.html#class-Regexp-label-Regexp-23match+method

[1] https://superuser.com/questions/129519/which-file-systems-support-newlines-in-file-names

Reply:
Why should the filename contain newline in the first place. This code is reused and has worked fine for some time.

Reply:
Well, reason is security. It is not critical for me, but in general for security reasons it is recommented to use `\A\z` instead `^$` in ruby. The second one worked, but can be tricked to not do what expected.

Reply:
OK, then I suggest you create a bug report so that is also gets fixed in yast2-storage-ng.